In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [3]:
train=pd.read_csv('train.csv')
test=pd.read_csv('test.csv')

In [3]:
train.head()

Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000.0       0      3       0     1       1     0    NaN   
1  ID_f29eb3ddd  135000.0       0      4       0     1       1     1    1.0   
2  ID_68de51c94       NaN       0      8       0     1       1     0    NaN   
3  ID_d671db89c  180000.0       0      5       0     1       1     1    1.0   
4  ID_d56d6f5f5  180000.0       0      5       0     1       1     1    1.0   

   r4h1  ...  SQBescolari  SQBage  SQBhogar_total  SQBedjefe  SQBhogar_nin  \
0     0  ...          100    1849               1        100             0   
1     0  ...          144    4489               1        144             0   
2     0  ...          121    8464               1          0             0   
3     0  ...           81     289              16        121             4   
4     0  ...          121    1369              16        121             4   

   SQBovercrowding  SQBdependency  SQBmeaned  agesq  Target  
0         1.000000            0.0      100.0   1849       4  
1         1.000000           64.0      144.0   4489       4  
2         0.250000           64.0      121.0   8464       4  
3         1.777778            1.0      121.0    289       4  
4         1.777778            1.0      121.0   1369       4  

[5 rows x 143 columns]

In [5]:
pd.set_option('display.max_columns',None)

In [6]:
train.head()

Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000.0       0      3       0     1       1     0    NaN   
1  ID_f29eb3ddd  135000.0       0      4       0     1       1     1    1.0   
2  ID_68de51c94       NaN       0      8       0     1       1     0    NaN   
3  ID_d671db89c  180000.0       0      5       0     1       1     1    1.0   
4  ID_d56d6f5f5  180000.0       0      5       0     1       1     1    1.0   

   r4h1  r4h2  r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  \
0     0     1     1     0     0     0     0     1     1       1       1   
1     0     1     1     0     0     0     0     1     1       1       1   
2     0     0     0     0     1     1     0     1     1       1       1   
3     0     2     2     1     1     2     1     3     4       4       4   
4     0     2     2     1     1     2     1     3     4       4       4   

   escolari  rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  \
0        10      NaN       1            1            0          0         0   
1        12      NaN       1            0            0          0         0   
2        11      NaN       1            0            0          0         0   
3         9      1.0       4            1            0          0         0   
4        11      NaN       4            1            0          0         0   

   paredmad  paredzinc  paredfibras  paredother  pisomoscer  pisocemento  \
0         0          0            0           0           1            0   
1         1          0            0           0           0            0   
2         1          0            0           0           1            0   
3         0          0            0           0           1            0   
4         0          0            0           0           1            0   

   pisoother  pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  \
0          0          0            0           0          0               1   
1          0          0            0           1          1               0   
2          0          0            0           0          1               0   
3          0          0            0           0          1               0   
4          0          0            0           0          1               0   

   techocane  techootro  cielorazo  abastaguadentro  abastaguafuera  \
0          0          0          1                1               0   
1          0          0          1                1               0   
2          0          0          1                1               0   
3          0          0          1                1               0   
4          0          0          1                1               0   

   abastaguano  public  planpri  noelec  coopele  sanitario1  sanitario2  \
0            0       1        0       0        0           0           1   
1            0       1        0       0        0           0           1   
2            0       1        0       0        0           0           1   
3            0       1        0       0        0           0           1   
4            0       1        0       0        0           0           1   

   sanitario3  sanitario5  sanitario6  energcocinar1  energcocinar2  \
0           0           0           0              0              0   
1           0           0           0              0              1   
2           0           0           0              0              1   
3           0           0           0              0              1   
4           0           0           0              0              1   

   energcocinar3  energcocinar4  elimbasu1  elimbasu2  elimbasu3  elimbasu4  \
0              1              0          1          0          0          0   
1              0              0          1          0          0          0   
2              0              0          1          0          0          0   
3              0              0          1          0

In [8]:
test.head()

Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_2f6873615       NaN       0      5       0     1       1     0    NaN   
1  ID_1c78846d2       NaN       0      5       0     1       1     0    NaN   
2  ID_e5442cf6a       NaN       0      5       0     1       1     0    NaN   
3  ID_a8db26a79       NaN       0     14       0     1       1     1    1.0   
4  ID_a62966799  175000.0       0      4       0     1       1     1    1.0   

   r4h1  r4h2  r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  \
0     1     1     2     0     1     1     1     2     3       3       3   
1     1     1     2     0     1     1     1     2     3       3       3   
2     1     1     2     0     1     1     1     2     3       3       3   
3     0     1     1     0     0     0     0     1     1       1       1   
4     0     0     0     0     1     1     0     1     1       1       1   

   escolari  rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  \
0         0      NaN       3            1            0          0         0   
1        16      NaN       3            1            0          0         0   
2        17      NaN       3            1            0          0         0   
3        16      NaN       1            1            0          0         0   
4        11      NaN       1            1            0          0         0   

   paredmad  paredzinc  paredfibras  paredother  pisomoscer  pisocemento  \
0         0          0            0           0           1            0   
1         0          0            0           0           1            0   
2         0          0            0           0           1            0   
3         0          0            0           0           1            0   
4         0          0            0           0           1            0   

   pisoother  pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  \
0          0          0            0           0          1               0   
1          0          0            0           0          1               0   
2          0          0            0           0          1               0   
3          0          0            0           0          1               0   
4          0          0            0           0          1               0   

   techocane  techootro  cielorazo  abastaguadentro  abastaguafuera  \
0          0          0          1                1               0   
1          0          0          1                1               0   
2          0          0          1                1               0   
3          0          0          1                1               0   
4          0          0          1                1               0   

   abastaguano  public  planpri  noelec  coopele  sanitario1  sanitario2  \
0            0       1        0       0        0           0           1   
1            0       1        0       0        0           0           1   
2            0       1        0       0        0           0           1   
3            0       1        0       0        0           0           1   
4            0       1        0       0        0           0           1   

   sanitario3  sanitario5  sanitario6  energcocinar1  energcocinar2  \
0           0           0           0              0              0   
1           0           0           0              0              0   
2           0           0           0              0              0   
3           0           0           0              0              1   
4           0           0           0              0              1   

   energcocinar3  energcocinar4  elimbasu1  elimbasu2  elimbasu3  elimbasu4  \
0              1              0          1          0          0          0   
1              1              0          1          0          0          0   
2              1              0          1          0          0          0   
3              0              0          1          0

In [9]:
train.describe()

v2a1       hacdor        rooms       hacapo         v14a  \
count  2.697000e+03  9557.000000  9557.000000  9557.000000  9557.000000   
mean   1.652316e+05     0.038087     4.955530     0.023648     0.994768   
std    1.504571e+05     0.191417     1.468381     0.151957     0.072145   
min    0.000000e+00     0.000000     1.000000     0.000000     0.000000   
25%    8.000000e+04     0.000000     4.000000     0.000000     1.000000   
50%    1.300000e+05     0.000000     5.000000     0.000000     1.000000   
75%    2.000000e+05     0.000000     6.000000     0.000000     1.000000   
max    2.353477e+06     1.000000    11.000000     1.000000     1.000000   

            refrig         v18q        v18q1         r4h1         r4h2  \
count  9557.000000  9557.000000  2215.000000  9557.000000  9557.000000   
mean      0.957623     0.231767     1.404063     0.385895     1.559171   
std       0.201459     0.421983     0.763131     0.680779     1.036574   
min       0.000000     0.000000     1.000000     0.000000     0.000000   
25%       1.000000     0.000000     1.000000     0.000000     1.000000   
50%       1.000000     0.000000     1.000000     0.000000     1.000000   
75%       1.000000     0.000000     2.000000     1.000000     2.000000   
max       1.000000     1.000000     6.000000     5.000000     8.000000   

              r4h3         r4m1         r4m2         r4m3         r4t1  \
count  9557.000000  9557.000000  9557.000000  9557.000000  9557.000000   
mean      1.945066     0.399184     1.661714     2.060898     0.785079   
std       1.188852     0.692460     0.933052     1.206172     1.047559   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       1.000000     0.000000     1.000000     1.000000     0.000000   
50%       2.000000     0.000000     1.000000     2.000000     0.000000   
75%       3.000000     1.000000     2.000000     3.000000     1.000000   
max       8.000000     6.000000     6.000000     8.000000     7.000000   

              r4t2         r4t3       tamhog       tamviv     escolari  \
count  9557.000000  9557.000000  9557.000000  9557.000000  9557.000000   
mean      3.220885     4.005964     3.999058     4.094590     7.200272   
std       1.440995     1.771202     1.772216     1.876428     4.730877   
min       1.000000     1.000000     1.000000     1.000000     0.000000   
25%       2.000000     3.000000     3.000000     3.000000     4.000000   
50%       3.000000     4.000000     4.000000     4.000000     6.000000   
75%       4.000000     5.000000     5.000000     5.000000    11.000000   
max      11.000000    13.000000    13.000000    15.000000    21.000000   

           rez_esc       hhsize  paredblolad  paredzocalo    paredpreb  \
count  1629.000000  9557.000000  9557.000000  9557.000000  9557.000000   
mean      0.459791     3.999058     0.594015     0.077744     0.188030   
std       0.946550     1.772216     0.491107     0.267782     0.390756   
min       0.000000     1.000000     0.000000     0.000000     0.000000   
25%       0.000000     3.000000     0.000000     0.000000     0.000000   
50%       0.000000     4.000000     1.000000     0.000000     0.000000   
75%       1.000000     5.000000     1.000000     0.000000     0.000000   
max       5.000000    13.000000     1.000000     1.000000     1.000000   

          pareddes     paredmad    paredzinc  paredfibras   paredother  \
count  9557.000000  9557.000000  9557.000000  9557.000000  9557.000000   
mean      0.008580     0.115622     0.013079     0.001465     0.001465   
std       0.092235     0.319788     0.113621     0.038248     0.038248   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

    

In [10]:
train['idhogar'].unique().size

2988

In [11]:
gp=train.groupby('idhogar')

In [12]:
dif_family=[]
for i in train['idhogar'].unique():
    dif_family.append(i)

In [13]:
id_fami={}
for i in dif_family:
    id_fami.update({i:gp.get_group(i)[['Id','Target']]})

In [14]:
lis={}

In [15]:
for i in dif_family:
    for j in id_fami[i]['Target']:
        lis.update({i:j})

In [17]:
df=pd.DataFrame()

In [18]:
df['Family']=lis.keys()
df['Target']=lis.values()

In [19]:
df

Family  Target
0     21eb7fcc1       4
1     0e5d7a658       4
2     2c7317ea8       4
3     2b58d945f       4
4     d6dae86b7       4
...         ...     ...
2983  9bbf7c6ca       2
2984  e87e70c06       2
2985  a8eeafc29       4
2986  212db6f6c       2
2987  d6c086aa3       2

[2988 rows x 2 columns]

In [20]:
check_all_same=[]

In [21]:
ra={}

In [22]:
for i in dif_family:
    for j in gp.get_group(i)['Id'] :
        k=train['Target'][train['Id']==j].iloc[0]
        if df['Target'][df['Family']==i].iloc[0]!=k:
            ra.update({i:"No"})
        else:
            continue

In [23]:
for i in ra.keys():
    check_all_same.append(i)

In [26]:
len(check_all_same)

85

There are 2988 houses in Latin America.
out of 2988, all members of 2903 houses on the same poverty level

members of 85 houses have diferent poverty level


In [36]:


hdli=[]
hd={}

In [37]:
for i in dif_family:
    for j in gp.get_group(i)['Id'] :
        k=train['parentesco1'][train['Id']==j].iloc[0]
        hdli.append(k)
    if 1 not in hdli:
        hd.update({i:"No"})
    hdli.clear()
    

In [38]:
hd

{'09b195e7a': 'No',
 '896fe6d3e': 'No',
 '61c10e099': 'No',
 '374ca5a19': 'No',
 'bfd5067c2': 'No',
 '1367ab31d': 'No',
 '6b1b2405f': 'No',
 'f2bfa75c4': 'No',
 '03c6bdf85': 'No',
 'ad687ad89': 'No',
 'b1f4d89d7': 'No',
 'c0c8a5013': 'No',
 'a0812ef17': 'No',
 'd363d9183': 'No',
 '1bc617b23': 'No'}

In [39]:
without_family_head=list(hd.keys())

In [40]:
without_family_head

['09b195e7a',
 '896fe6d3e',
 '61c10e099',
 '374ca5a19',
 'bfd5067c2',
 '1367ab31d',
 '6b1b2405f',
 'f2bfa75c4',
 '03c6bdf85',
 'ad687ad89',
 'b1f4d89d7',
 'c0c8a5013',
 'a0812ef17',
 'd363d9183',
 '1bc617b23']

In [41]:
len(without_family_head)

15

There 15 houses without a family head.

In [43]:
for i in without_family_head:
    print(train[['Target','Id','idhogar']][train['idhogar']==i])

      Target            Id    idhogar
4935       3  ID_717302050  09b195e7a
      Target            Id    idhogar
4975       3  ID_ff6f13458  896fe6d3e
      Target            Id    idhogar
5391       3  ID_64c9eabac  61c10e099
      Target            Id    idhogar
5396       3  ID_8948b1f91  374ca5a19
      Target            Id    idhogar
6443       3  ID_8ddb34d2c  bfd5067c2
6444       3  ID_03e2f23b0  bfd5067c2
      Target            Id    idhogar
7086       3  ID_ba2ac56bb  1367ab31d
      Target            Id    idhogar
7438       4  ID_7ba374a80  6b1b2405f
7439       4  ID_1bca5b5e5  6b1b2405f
7440       4  ID_622be4200  6b1b2405f
      Target            Id    idhogar
7461       3  ID_1f91daf92  f2bfa75c4
7462       3  ID_686d2639f  f2bfa75c4
7463       3  ID_b84170356  f2bfa75c4
      Target            Id    idhogar
7705       3  ID_c7c038825  03c6bdf85
7706       3  ID_ebef0991f  03c6bdf85
      Target            Id    idhogar
7756       3  ID_f81c03ca9  ad687ad89
      Target

In [45]:
def family(fa):
    for j in gp.get_group(fa)['Id']:
        k=train['parentesco1'][train['Id']==j].iloc[0]
        if k==1:
            a=train['Target'][train['Id']==j].iloc[0]
            return(a)
    

In [44]:
newTar=[]

In [46]:
for i in train['idhogar']:
    newTar.append(family(i))

In [75]:
train['New Target']=newTar

In [68]:
train.head()

Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000.0       0      3       0     1       1     0    NaN   
1  ID_f29eb3ddd  135000.0       0      4       0     1       1     1    1.0   
2  ID_68de51c94       NaN       0      8       0     1       1     0    NaN   
3  ID_d671db89c  180000.0       0      5       0     1       1     1    1.0   
4  ID_d56d6f5f5  180000.0       0      5       0     1       1     1    1.0   

   r4h1  r4h2  r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  \
0     0     1     1     0     0     0     0     1     1       1       1   
1     0     1     1     0     0     0     0     1     1       1       1   
2     0     0     0     0     1     1     0     1     1       1       1   
3     0     2     2     1     1     2     1     3     4       4       4   
4     0     2     2     1     1     2     1     3     4       4       4   

   escolari  rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  \
0        10      NaN       1            1            0          0         0   
1        12      NaN       1            0            0          0         0   
2        11      NaN       1            0            0          0         0   
3         9      1.0       4            1            0          0         0   
4        11      NaN       4            1            0          0         0   

   paredmad  paredzinc  paredfibras  paredother  pisomoscer  pisocemento  \
0         0          0            0           0           1            0   
1         1          0            0           0           0            0   
2         1          0            0           0           1            0   
3         0          0            0           0           1            0   
4         0          0            0           0           1            0   

   pisoother  pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  \
0          0          0            0           0          0               1   
1          0          0            0           1          1               0   
2          0          0            0           0          1               0   
3          0          0            0           0          1               0   
4          0          0            0           0          1               0   

   techocane  techootro  cielorazo  abastaguadentro  abastaguafuera  \
0          0          0          1                1               0   
1          0          0          1                1               0   
2          0          0          1                1               0   
3          0          0          1                1               0   
4          0          0          1                1               0   

   abastaguano  public  planpri  noelec  coopele  sanitario1  sanitario2  \
0            0       1        0       0        0           0           1   
1            0       1        0       0        0           0           1   
2            0       1        0       0        0           0           1   
3            0       1        0       0        0           0           1   
4            0       1        0       0        0           0           1   

   sanitario3  sanitario5  sanitario6  energcocinar1  energcocinar2  \
0           0           0           0              0              0   
1           0           0           0              0              1   
2           0           0           0              0              1   
3           0           0           0              0              1   
4           0           0           0              0              1   

   energcocinar3  energcocinar4  elimbasu1  elimbasu2  elimbasu3  elimbasu4  \
0              1              0          1          0          0          0   
1              0              0          1          0          0          0   
2              0              0          1          0          0          0   
3              0              0          1          0

In [78]:
mis=train[['Target','New Target','idhogar']][train['Target']!=train['New Target']]

In [79]:
mis.shape

(163, 3)

In [80]:
mis['New Target'].isnull().sum()

23

In [81]:
mis['New Target'][mis['New Target'].isnull()]=mis['Target'][mis['New Target'].isnull()]

In [82]:
mis['New Target'].isnull().sum()

0

In [83]:
mis['New Target']

283     1.0
284     1.0
290     2.0
322     2.0
410     3.0
       ... 
9497    3.0
9536    1.0
9537    1.0
9538    1.0
9539    1.0
Name: New Target, Length: 163, dtype: float64

In [85]:
train[['Target','New Target','idhogar']][train['Target']!=train['New Target']]

Target  New Target    idhogar
283        2         1.0  4b6077882
284        2         1.0  4b6077882
290        1         2.0  6833ac5dc
322        1         2.0  43b9c83e5
410        2         3.0  5c3f7725d
...      ...         ...        ...
9497       3         NaN  1bc617b23
9536       2         1.0  9bbf7c6ca
9537       2         1.0  9bbf7c6ca
9538       2         1.0  9bbf7c6ca
9539       2         1.0  9bbf7c6ca

[163 rows x 3 columns]

In [88]:
train['New Target'][train['Target']!=train['New Target']]=mis['New Target']

In [89]:
train[['Target','New Target','idhogar']][train['Target']!=train['New Target']]

Target  New Target    idhogar
283        2         1.0  4b6077882
284        2         1.0  4b6077882
290        1         2.0  6833ac5dc
322        1         2.0  43b9c83e5
410        2         3.0  5c3f7725d
...      ...         ...        ...
9475       1         2.0  c7ce4e30c
9536       2         1.0  9bbf7c6ca
9537       2         1.0  9bbf7c6ca
9538       2         1.0  9bbf7c6ca
9539       2         1.0  9bbf7c6ca

[140 rows x 3 columns]

In [91]:
train=train.drop('Target',axis=1)

In [92]:
train.head()

Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000.0       0      3       0     1       1     0    NaN   
1  ID_f29eb3ddd  135000.0       0      4       0     1       1     1    1.0   
2  ID_68de51c94       NaN       0      8       0     1       1     0    NaN   
3  ID_d671db89c  180000.0       0      5       0     1       1     1    1.0   
4  ID_d56d6f5f5  180000.0       0      5       0     1       1     1    1.0   

   r4h1  r4h2  r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  \
0     0     1     1     0     0     0     0     1     1       1       1   
1     0     1     1     0     0     0     0     1     1       1       1   
2     0     0     0     0     1     1     0     1     1       1       1   
3     0     2     2     1     1     2     1     3     4       4       4   
4     0     2     2     1     1     2     1     3     4       4       4   

   escolari  rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  \
0        10      NaN       1            1            0          0         0   
1        12      NaN       1            0            0          0         0   
2        11      NaN       1            0            0          0         0   
3         9      1.0       4            1            0          0         0   
4        11      NaN       4            1            0          0         0   

   paredmad  paredzinc  paredfibras  paredother  pisomoscer  pisocemento  \
0         0          0            0           0           1            0   
1         1          0            0           0           0            0   
2         1          0            0           0           1            0   
3         0          0            0           0           1            0   
4         0          0            0           0           1            0   

   pisoother  pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  \
0          0          0            0           0          0               1   
1          0          0            0           1          1               0   
2          0          0            0           0          1               0   
3          0          0            0           0          1               0   
4          0          0            0           0          1               0   

   techocane  techootro  cielorazo  abastaguadentro  abastaguafuera  \
0          0          0          1                1               0   
1          0          0          1                1               0   
2          0          0          1                1               0   
3          0          0          1                1               0   
4          0          0          1                1               0   

   abastaguano  public  planpri  noelec  coopele  sanitario1  sanitario2  \
0            0       1        0       0        0           0           1   
1            0       1        0       0        0           0           1   
2            0       1        0       0        0           0           1   
3            0       1        0       0        0           0           1   
4            0       1        0       0        0           0           1   

   sanitario3  sanitario5  sanitario6  energcocinar1  energcocinar2  \
0           0           0           0              0              0   
1           0           0           0              0              1   
2           0           0           0              0              1   
3           0           0           0              0              1   
4           0           0           0              0              1   

   energcocinar3  energcocinar4  elimbasu1  elimbasu2  elimbasu3  elimbasu4  \
0              1              0          1          0          0          0   
1              0              0          1          0          0          0   
2              0              0          1          0          0          0   
3              0              0          1          0

In [95]:
train.isnull().sum()

Id                    0
v2a1               6860
hacdor                0
rooms                 0
hacapo                0
                   ... 
SQBovercrowding       0
SQBdependency         0
SQBmeaned             5
agesq                 0
New Target            0
Length: 143, dtype: int64

In [96]:
for i in train.columns:
    if(train[i].isnull().any()):
        print(i,train[i].isnull().sum())

v2a1 6860
v18q1 7342
rez_esc 7928
meaneduc 5
SQBmeaned 5


In [104]:
train['v2a1'][train['v2a1'].isnull()]

2      NaN
13     NaN
14     NaN
26     NaN
32     NaN
        ..
9542   NaN
9543   NaN
9544   NaN
9550   NaN
9551   NaN
Name: v2a1, Length: 6860, dtype: float64

In [105]:
for j in gp.get_group('2b58d945f')['Id']:
    print(train['v2a1'][train['Id']==j])

3    180000.0
Name: v2a1, dtype: float64
4    180000.0
Name: v2a1, dtype: float64
5    180000.0
Name: v2a1, dtype: float64
6    180000.0
Name: v2a1, dtype: float64


In [106]:
train['v2a1']=train['v2a1'].fillna(0)

In [107]:
train.head()

Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000.0       0      3       0     1       1     0    NaN   
1  ID_f29eb3ddd  135000.0       0      4       0     1       1     1    1.0   
2  ID_68de51c94       0.0       0      8       0     1       1     0    NaN   
3  ID_d671db89c  180000.0       0      5       0     1       1     1    1.0   
4  ID_d56d6f5f5  180000.0       0      5       0     1       1     1    1.0   

   r4h1  r4h2  r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  \
0     0     1     1     0     0     0     0     1     1       1       1   
1     0     1     1     0     0     0     0     1     1       1       1   
2     0     0     0     0     1     1     0     1     1       1       1   
3     0     2     2     1     1     2     1     3     4       4       4   
4     0     2     2     1     1     2     1     3     4       4       4   

   escolari  rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  \
0        10      NaN       1            1            0          0         0   
1        12      NaN       1            0            0          0         0   
2        11      NaN       1            0            0          0         0   
3         9      1.0       4            1            0          0         0   
4        11      NaN       4            1            0          0         0   

   paredmad  paredzinc  paredfibras  paredother  pisomoscer  pisocemento  \
0         0          0            0           0           1            0   
1         1          0            0           0           0            0   
2         1          0            0           0           1            0   
3         0          0            0           0           1            0   
4         0          0            0           0           1            0   

   pisoother  pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  \
0          0          0            0           0          0               1   
1          0          0            0           1          1               0   
2          0          0            0           0          1               0   
3          0          0            0           0          1               0   
4          0          0            0           0          1               0   

   techocane  techootro  cielorazo  abastaguadentro  abastaguafuera  \
0          0          0          1                1               0   
1          0          0          1                1               0   
2          0          0          1                1               0   
3          0          0          1                1               0   
4          0          0          1                1               0   

   abastaguano  public  planpri  noelec  coopele  sanitario1  sanitario2  \
0            0       1        0       0        0           0           1   
1            0       1        0       0        0           0           1   
2            0       1        0       0        0           0           1   
3            0       1        0       0        0           0           1   
4            0       1        0       0        0           0           1   

   sanitario3  sanitario5  sanitario6  energcocinar1  energcocinar2  \
0           0           0           0              0              0   
1           0           0           0              0              1   
2           0           0           0              0              1   
3           0           0           0              0              1   
4           0           0           0              0              1   

   energcocinar3  energcocinar4  elimbasu1  elimbasu2  elimbasu3  elimbasu4  \
0              1              0          1          0          0          0   
1              0              0          1          0          0          0   
2              0              0          1          0          0          0   
3              0              0          1          0

In [109]:
train['v2a1']=train.v2a1.astype(int)

In [112]:
train['New Target']=train['New Target'].astype(int)

In [113]:
train.head()

Id    v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000       0      3       0     1       1     0    NaN   
1  ID_f29eb3ddd  135000       0      4       0     1       1     1    1.0   
2  ID_68de51c94       0       0      8       0     1       1     0    NaN   
3  ID_d671db89c  180000       0      5       0     1       1     1    1.0   
4  ID_d56d6f5f5  180000       0      5       0     1       1     1    1.0   

   r4h1  r4h2  r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  \
0     0     1     1     0     0     0     0     1     1       1       1   
1     0     1     1     0     0     0     0     1     1       1       1   
2     0     0     0     0     1     1     0     1     1       1       1   
3     0     2     2     1     1     2     1     3     4       4       4   
4     0     2     2     1     1     2     1     3     4       4       4   

   escolari  rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  \
0        10      NaN       1            1            0          0         0   
1        12      NaN       1            0            0          0         0   
2        11      NaN       1            0            0          0         0   
3         9      1.0       4            1            0          0         0   
4        11      NaN       4            1            0          0         0   

   paredmad  paredzinc  paredfibras  paredother  pisomoscer  pisocemento  \
0         0          0            0           0           1            0   
1         1          0            0           0           0            0   
2         1          0            0           0           1            0   
3         0          0            0           0           1            0   
4         0          0            0           0           1            0   

   pisoother  pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  \
0          0          0            0           0          0               1   
1          0          0            0           1          1               0   
2          0          0            0           0          1               0   
3          0          0            0           0          1               0   
4          0          0            0           0          1               0   

   techocane  techootro  cielorazo  abastaguadentro  abastaguafuera  \
0          0          0          1                1               0   
1          0          0          1                1               0   
2          0          0          1                1               0   
3          0          0          1                1               0   
4          0          0          1                1               0   

   abastaguano  public  planpri  noelec  coopele  sanitario1  sanitario2  \
0            0       1        0       0        0           0           1   
1            0       1        0       0        0           0           1   
2            0       1        0       0        0           0           1   
3            0       1        0       0        0           0           1   
4            0       1        0       0        0           0           1   

   sanitario3  sanitario5  sanitario6  energcocinar1  energcocinar2  \
0           0           0           0              0              0   
1           0           0           0              0              1   
2           0           0           0              0              1   
3           0           0           0              0              1   
4           0           0           0              0              1   

   energcocinar3  energcocinar4  elimbasu1  elimbasu2  elimbasu3  elimbasu4  \
0              1              0          1          0          0          0   
1              0              0          1          0          0          0   
2              0              0          1          0          0          0   
3              0              0          1          0          0 

In [118]:
for j in gp.get_group('2b58d945f')['Id']:
    print(train['v18q1'][train['Id']==j])

3    1.0
Name: v18q1, dtype: float64
4    1.0
Name: v18q1, dtype: float64
5    1.0
Name: v18q1, dtype: float64
6    1.0
Name: v18q1, dtype: float64


In [125]:
train['v18q1']=train['v18q1'].fillna(0)

In [126]:
train['v18q1']=train['v18q1'].astype(int)
train.head()

Id    v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000       0      3       0     1       1     0      0   
1  ID_f29eb3ddd  135000       0      4       0     1       1     1      1   
2  ID_68de51c94       0       0      8       0     1       1     0      0   
3  ID_d671db89c  180000       0      5       0     1       1     1      1   
4  ID_d56d6f5f5  180000       0      5       0     1       1     1      1   

   r4h1  r4h2  r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  \
0     0     1     1     0     0     0     0     1     1       1       1   
1     0     1     1     0     0     0     0     1     1       1       1   
2     0     0     0     0     1     1     0     1     1       1       1   
3     0     2     2     1     1     2     1     3     4       4       4   
4     0     2     2     1     1     2     1     3     4       4       4   

   escolari  rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  \
0        10      NaN       1            1            0          0         0   
1        12      NaN       1            0            0          0         0   
2        11      NaN       1            0            0          0         0   
3         9      1.0       4            1            0          0         0   
4        11      NaN       4            1            0          0         0   

   paredmad  paredzinc  paredfibras  paredother  pisomoscer  pisocemento  \
0         0          0            0           0           1            0   
1         1          0            0           0           0            0   
2         1          0            0           0           1            0   
3         0          0            0           0           1            0   
4         0          0            0           0           1            0   

   pisoother  pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  \
0          0          0            0           0          0               1   
1          0          0            0           1          1               0   
2          0          0            0           0          1               0   
3          0          0            0           0          1               0   
4          0          0            0           0          1               0   

   techocane  techootro  cielorazo  abastaguadentro  abastaguafuera  \
0          0          0          1                1               0   
1          0          0          1                1               0   
2          0          0          1                1               0   
3          0          0          1                1               0   
4          0          0          1                1               0   

   abastaguano  public  planpri  noelec  coopele  sanitario1  sanitario2  \
0            0       1        0       0        0           0           1   
1            0       1        0       0        0           0           1   
2            0       1        0       0        0           0           1   
3            0       1        0       0        0           0           1   
4            0       1        0       0        0           0           1   

   sanitario3  sanitario5  sanitario6  energcocinar1  energcocinar2  \
0           0           0           0              0              0   
1           0           0           0              0              1   
2           0           0           0              0              1   
3           0           0           0              0              1   
4           0           0           0              0              1   

   energcocinar3  energcocinar4  elimbasu1  elimbasu2  elimbasu3  elimbasu4  \
0              1              0          1          0          0          0   
1              0              0          1          0          0          0   
2              0              0          1          0          0          0   
3              0              0          1          0          0 

In [133]:
train[train['rez_esc'].isnull()==False]

Id    v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
3     ID_d671db89c  180000       0      5       0     1       1     1      1   
6     ID_e9e0c1100  180000       0      5       0     1       1     1      1   
7     ID_3e04e571e  130000       1      2       0     1       1     0      0   
10    ID_db44f5c59  130000       1      2       0     1       1     0      0   
38    ID_0f01e18e9   80000       0      4       0     1       1     0      0   
...            ...     ...     ...    ...     ...   ...     ...   ...    ...   
9516  ID_9b58089fb       0       0      8       0     1       1     1      1   
9533  ID_5384717d3   70000       0      4       0     1       1     0      0   
9536  ID_d20bd7576       0       0      4       0     1       0     0      0   
9542  ID_fc386a944       0       0      3       0     0       0     0      0   
9543  ID_2f8268634       0       0      3       0     0       0     0      0   

      r4h1  r4h2  r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  \
3        0     2     2     1     1     2     1     3     4       4       4   
6        0     2     2     1     1     2     1     3     4       4       4   
7        0     1     1     2     1     3     2     2     4       4       4   
10       0     1     1     2     1     3     2     2     4       4       4   
38       0     2     2     0     1     1     0     3     3       3       3   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...     ...     ...   
9516     0     2     2     0     1     1     0     3     3       3       3   
9533     0     2     2     0     1     1     0     3     3       3       3   
9536     1     1     2     2     1     3     3     2     5       5       5   
9542     2     2     4     0     1     1     2     3     5       5       5   
9543     2     2     4     0     1     1     2     3     5       5       5   

      escolari  rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  \
3            9      1.0       4            1            0          0   
6            2      0.0       4            1            0          0   
7            0      0.0       4            1            0          0   
10           3      1.0       4            1            0          0   
38           8      0.0       3            0            0          0   
...        ...      ...     ...          ...          ...        ...   
9516         6      0.0       3            0            0          0   
9533        10      0.0       3            0            1          0   
9536         0      1.0       5            0            1          0   
9542         3      0.0       5            0            0          0   
9543         6      0.0       5            0            0          0   

      pareddes  paredmad  paredzinc  paredfibras  paredother  pisomoscer  \
3            0         0          0            0           0           1   
6            0         0          0            0           0           1   
7            0         0          0            0           0           1   
10           0         0          0            0           0           1   
38           0         1          0            0           0           0   
...        ...       ...        ...          ...         ...         ...   
9516         0         1          0            0           0           0   
9533         0         0          0            0           0           0   
9536         0         0          0            0           0           0   
9542         0         1          0            0           0           0   
9543         0         1          0            0           0           0   

      pisocemento  pisoother  pisonatur  pisonotiene  pisomadera  techozinc  \
3               0          0          0            0           0          1   
6               0          0          0            0           0          1   
7               0          0          0            0           0          1   
10              0       

In [134]:
train['rez_esc']=train['rez_esc'].fillna(0)

In [147]:
train.rez_esc=train.rez_esc.astype(int)

In [138]:
train['meaneduc']=train['meaneduc'].fillna(train['meaneduc'].mean())

In [139]:
train['meaneduc']=train['meaneduc'].astype(int)

In [140]:
#SQBmeaned
train['SQBmeaned']=train['SQBmeaned'].fillna(train['SQBmeaned'].mean())

In [141]:
train['SQBmeaned']=train['SQBmeaned'].astype(int)

In [162]:
for i in train.columns:
    if (train[i].dtype !='int64'):
        print(i)

Id
v2a1
v18q1
rez_esc
idhogar
dependency
edjefe
edjefa
meaneduc
overcrowding
SQBovercrowding
SQBdependency
SQBmeaned
New Target


In [173]:
train.Id.count()

9557

In [188]:
id_map={}

In [189]:
for i in range(0,9557):
    id_map.update({train['Id'].iloc[i]:i})

In [182]:
train['Id']=id_map.values()

In [183]:
train.head()

Id    v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  r4h1  r4h2  \
0   0  190000       0      3       0     1       1     0      0     0     1   
1   1  135000       0      4       0     1       1     1      1     0     1   
2   2       0       0      8       0     1       1     0      0     0     0   
3   3  180000       0      5       0     1       1     1      1     0     2   
4   4  180000       0      5       0     1       1     1      1     0     2   

   r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  escolari  \
0     1     0     0     0     0     1     1       1       1        10   
1     1     0     0     0     0     1     1       1       1        12   
2     0     0     1     1     0     1     1       1       1        11   
3     2     1     1     2     1     3     4       4       4         9   
4     2     1     1     2     1     3     4       4       4        11   

   rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  paredmad  \
0        0       1            1            0          0         0         0   
1        0       1            0            0          0         0         1   
2        0       1            0            0          0         0         1   
3        1       4            1            0          0         0         0   
4        0       4            1            0          0         0         0   

   paredzinc  paredfibras  paredother  pisomoscer  pisocemento  pisoother  \
0          0            0           0           1            0          0   
1          0            0           0           0            0          0   
2          0            0           0           1            0          0   
3          0            0           0           1            0          0   
4          0            0           0           1            0          0   

   pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  techocane  \
0          0            0           0          0               1          0   
1          0            0           1          1               0          0   
2          0            0           0          1               0          0   
3          0            0           0          1               0          0   
4          0            0           0          1               0          0   

   techootro  cielorazo  abastaguadentro  abastaguafuera  abastaguano  public  \
0          0          1                1               0            0       1   
1          0          1                1               0            0       1   
2          0          1                1               0            0       1   
3          0          1                1               0            0       1   
4          0          1                1               0            0       1   

   planpri  noelec  coopele  sanitario1  sanitario2  sanitario3  sanitario5  \
0        0       0        0           0           1           0           0   
1        0       0        0           0           1           0           0   
2        0       0        0           0           1           0           0   
3        0       0        0           0           1           0           0   
4        0       0        0           0           1           0           0   

   sanitario6  energcocinar1  energcocinar2  energcocinar3  energcocinar4  \
0           0              0              0              1              0   
1           0              0              1              0              0   
2           0              0              1              0              0   
3           0              0              1              0              0   
4           0              0              1              0              0   

   elimbasu1  elimbasu2  elimbasu3  elimbasu4  elimbasu5  elimbasu6  epared1  \
0          1          0          0          0          0          0        0   
1          1          0          0          0          0          0        0   
2          

In [196]:
family_map={}
count=0

In [197]:
for i in range(0,9557):
    if train['idhogar'].iloc[i] not in family_map:
        family_map.update({train['idhogar'].iloc[i]:count})
        count+=1
    else:
        continue

In [203]:
def famap(nm):
    return(family_map[nm])

In [211]:
train['idhogar']=train['idhogar'].apply(famap)

In [212]:
train.head()

Id    v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  r4h1  r4h2  \
0   0  190000       0      3       0     1       1     0      0     0     1   
1   1  135000       0      4       0     1       1     1      1     0     1   
2   2       0       0      8       0     1       1     0      0     0     0   
3   3  180000       0      5       0     1       1     1      1     0     2   
4   4  180000       0      5       0     1       1     1      1     0     2   

   r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  escolari  \
0     1     0     0     0     0     1     1       1       1        10   
1     1     0     0     0     0     1     1       1       1        12   
2     0     0     1     1     0     1     1       1       1        11   
3     2     1     1     2     1     3     4       4       4         9   
4     2     1     1     2     1     3     4       4       4        11   

   rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  paredmad  \
0        0       1            1            0          0         0         0   
1        0       1            0            0          0         0         1   
2        0       1            0            0          0         0         1   
3        1       4            1            0          0         0         0   
4        0       4            1            0          0         0         0   

   paredzinc  paredfibras  paredother  pisomoscer  pisocemento  pisoother  \
0          0            0           0           1            0          0   
1          0            0           0           0            0          0   
2          0            0           0           1            0          0   
3          0            0           0           1            0          0   
4          0            0           0           1            0          0   

   pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  techocane  \
0          0            0           0          0               1          0   
1          0            0           1          1               0          0   
2          0            0           0          1               0          0   
3          0            0           0          1               0          0   
4          0            0           0          1               0          0   

   techootro  cielorazo  abastaguadentro  abastaguafuera  abastaguano  public  \
0          0          1                1               0            0       1   
1          0          1                1               0            0       1   
2          0          1                1               0            0       1   
3          0          1                1               0            0       1   
4          0          1                1               0            0       1   

   planpri  noelec  coopele  sanitario1  sanitario2  sanitario3  sanitario5  \
0        0       0        0           0           1           0           0   
1        0       0        0           0           1           0           0   
2        0       0        0           0           1           0           0   
3        0       0        0           0           1           0           0   
4        0       0        0           0           1           0           0   

   sanitario6  energcocinar1  energcocinar2  energcocinar3  energcocinar4  \
0           0              0              0              1              0   
1           0              0              1              0              0   
2           0              0              1              0              0   
3           0              0              1              0              0   
4           0              0              1              0              0   

   elimbasu1  elimbasu2  elimbasu3  elimbasu4  elimbasu5  elimbasu6  epared1  \
0          1          0          0          0          0          0        0   
1          1          0          0          0          0          0        0   
2          

In [213]:
def sno(nm):
    if nm=='no':
        return(0)
    else:
        return(1)

In [215]:
train[['dependency','edjefe','edjefa']]=train[['dependency','edjefe','edjefa']].applymap(sno)

In [216]:
train.head()

Id    v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  r4h1  r4h2  \
0   0  190000       0      3       0     1       1     0      0     0     1   
1   1  135000       0      4       0     1       1     1      1     0     1   
2   2       0       0      8       0     1       1     0      0     0     0   
3   3  180000       0      5       0     1       1     1      1     0     2   
4   4  180000       0      5       0     1       1     1      1     0     2   

   r4h3  r4m1  r4m2  r4m3  r4t1  r4t2  r4t3  tamhog  tamviv  escolari  \
0     1     0     0     0     0     1     1       1       1        10   
1     1     0     0     0     0     1     1       1       1        12   
2     0     0     1     1     0     1     1       1       1        11   
3     2     1     1     2     1     3     4       4       4         9   
4     2     1     1     2     1     3     4       4       4        11   

   rez_esc  hhsize  paredblolad  paredzocalo  paredpreb  pareddes  paredmad  \
0        0       1            1            0          0         0         0   
1        0       1            0            0          0         0         1   
2        0       1            0            0          0         0         1   
3        1       4            1            0          0         0         0   
4        0       4            1            0          0         0         0   

   paredzinc  paredfibras  paredother  pisomoscer  pisocemento  pisoother  \
0          0            0           0           1            0          0   
1          0            0           0           0            0          0   
2          0            0           0           1            0          0   
3          0            0           0           1            0          0   
4          0            0           0           1            0          0   

   pisonatur  pisonotiene  pisomadera  techozinc  techoentrepiso  techocane  \
0          0            0           0          0               1          0   
1          0            0           1          1               0          0   
2          0            0           0          1               0          0   
3          0            0           0          1               0          0   
4          0            0           0          1               0          0   

   techootro  cielorazo  abastaguadentro  abastaguafuera  abastaguano  public  \
0          0          1                1               0            0       1   
1          0          1                1               0            0       1   
2          0          1                1               0            0       1   
3          0          1                1               0            0       1   
4          0          1                1               0            0       1   

   planpri  noelec  coopele  sanitario1  sanitario2  sanitario3  sanitario5  \
0        0       0        0           0           1           0           0   
1        0       0        0           0           1           0           0   
2        0       0        0           0           1           0           0   
3        0       0        0           0           1           0           0   
4        0       0        0           0           1           0           0   

   sanitario6  energcocinar1  energcocinar2  energcocinar3  energcocinar4  \
0           0              0              0              1              0   
1           0              0              1              0              0   
2           0              0              1              0              0   
3           0              0              1              0              0   
4           0              0              1              0              0   

   elimbasu1  elimbasu2  elimbasu3  elimbasu4  elimbasu5  elimbasu6  epared1  \
0          1          0          0          0          0          0        0   
1          1          0          0          0          0          0        0   
2          

In [222]:
for i in train.columns:
    j=train[i].dtype
    print(i,j)

Id int32
v2a1 int32
hacdor int64
rooms int64
hacapo int64
v14a int64
refrig int64
v18q int64
v18q1 int32
r4h1 int64
r4h2 int64
r4h3 int64
r4m1 int64
r4m2 int64
r4m3 int64
r4t1 int64
r4t2 int64
r4t3 int64
tamhog int64
tamviv int64
escolari int64
rez_esc int32
hhsize int64
paredblolad int64
paredzocalo int64
paredpreb int64
pareddes int64
paredmad int64
paredzinc int64
paredfibras int64
paredother int64
pisomoscer int64
pisocemento int64
pisoother int64
pisonatur int64
pisonotiene int64
pisomadera int64
techozinc int64
techoentrepiso int64
techocane int64
techootro int64
cielorazo int64
abastaguadentro int64
abastaguafuera int64
abastaguano int64
public int64
planpri int64
noelec int64
coopele int64
sanitario1 int64
sanitario2 int64
sanitario3 int64
sanitario5 int64
sanitario6 int64
energcocinar1 int64
energcocinar2 int64
energcocinar3 int64
energcocinar4 int64
elimbasu1 int64
elimbasu2 int64
elimbasu3 int64
elimbasu4 int64
elimbasu5 int64
elimbasu6 int64
epared1 int64
epared2 int64
epar

In [163]:
from sklearn.ensemble import RandomForestClassifier

In [164]:
from sklearn.model_selection import train_test_split

In [227]:
X=train.drop(['New Target','Id'],axis=1)

In [218]:
y=train['New Target']

In [228]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33)

In [229]:
model=RandomForestClassifier(n_estimators=100,criterion='entropy',random_state=0)

In [230]:
model.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [237]:
from sklearn.metrics import confusion_matrix,accuracy_score

In [232]:
y_pred=model.predict(X_test)

In [234]:
confusion_matrix(y_test,y_pred)

array([[ 211,    5,    3,   46],
       [   2,  406,    0,   97],
       [   2,    7,  285,   81],
       [   0,   10,    1, 1998]], dtype=int64)

In [235]:
model.score(X_test,y_test)

0.9194673430564363

In [239]:
accuracy_score(y_pred,y_test)

0.9194673430564363